snuba/datasets/plans/partitioned_storage.py

```diff
@@ -0,0 +1,153 @@
+from abc import ABC, abstractmethod
+from typing import Optional, Sequence
+
+import sentry_sdk
+
+from snuba.clickhouse.query_dsl.accessors import get_object_ids_in_query_ast
+from snuba.clickhouse.translators.snuba.mapping import TranslationMappers
+from snuba.clusters.cluster import ClickhouseCluster, get_cluster
+from snuba.clusters.storage_sets import StorageSetKey
+from snuba.datasets.partitioning import (
+    is_storage_set_partitioned,
+    map_logical_partition_to_physical_partition,
+    map_org_id_to_logical_partition,
+)
+from snuba.datasets.plans.query_plan import (
+    ClickhouseQueryPlan,
+    ClickhouseQueryPlanBuilder,
+)
+from snuba.datasets.plans.single_storage import (
+    SimpleQueryPlanExecutionStrategy,
+    get_query_data_source,
+)
+from snuba.datasets.plans.translator.query import QueryTranslator
+from snuba.datasets.storage import ReadableStorage
+from snuba.query.logical import Query as LogicalQuery
+from snuba.query.processors.physical import ClickhouseQueryProcessor
+from snuba.query.processors.physical.conditions_enforcer import (
+    MandatoryConditionEnforcer,
+)
+from snuba.query.processors.physical.mandatory_condition_applier import (
+    MandatoryConditionApplier,
+)
+from snuba.query.query_settings import QuerySettings
+from snuba.util import with_span
+
+
+class StorageClusterSelector(ABC):
+    """
+    The component provided by a dataset and used at the beginning of the
+    execution of a query to pick the storage set a query should be executed
+    onto.
+    """
+
+    @abstractmethod
+    def select_cluster(
+        self, query: LogicalQuery, query_settings: QuerySettings
+    ) -> ClickhouseCluster:
+        raise NotImplementedError
+
+
+class ColumnBasedStoragePartitionSelector(StorageClusterSelector):
+    """
+    Storage partition selector for the generic metrics storage. This is needed
+    because the generic metrics storage can be partitioned and we would need to
+    know which partition to use for a specific query.
+    """
+
+    def __init__(
+        self, storage_set: StorageSetKey, partition_key_column_name: str
+    ) -> None:
+        self.storage_set = storage_set
+        self.partition_key_column_name = partition_key_column_name
+
+    def select_cluster(
+        self, query: LogicalQuery, query_settings: QuerySettings
+    ) -> ClickhouseCluster:
+        """
+        Selects the cluster to use for a query if the storage set is partitioned.
+        If the storage set is not partitioned, it returns the default cluster.
+        """
+        if not is_storage_set_partitioned(self.storage_set):
```

Comment:
@onewland - A storage may or may not be partitioned based on the environment we are running in (SaaS, ST, Self hosted). I have used https://github.com/getsentry/snuba/pull/3172/files#diff-4242bc3893c530314f2c6131401b0da4629e90cd6a24783a09273befb003c7daR28-R34 to make that decision. But I am not sure if that is the correct way to find out.

Reply:
I think that's right, thanks.